### PR TITLE
Implement Link and PartialLink text selectors for FindElement and related WebDriver commands

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23820)

### CODE REVIEW

File: components/webdriver_server/lib.rs [hunk 1]
```diff
@@ -1058,6 +1058,24 @@ impl Handler {
...
+                    false,
+                    sender,
+                );
+                self.browsing_context_script_command(cmd)?;
+            },
+            LocatorStrategy::PartialLinkText => {
+                let cmd = WebDriverScriptCommand::FindElementElementsLinkText(
+                    parameters.value.clone(),
+                    element.id.clone(),
+                    true,
```
Comment (jdm) [inline components/webdriver_server/lib.rs]: For all of these commits, we could reduce some of the duplication:
```rust
LocatorStrategy::LinkText | LocatorStrategy::PartialLinkText => {
    let cmd = WebDriverScriptCommand::FindElementElementsLinkText(
        parameters.value.clone(),
        element.id.clone(),
        parameters.using == LocatorStrategy::PartialLinkText,
        sender,
    );
    self.browsing_context_script_command(cmd)?;
}

File: components/script/webdriver_handlers.rs [hunk 2]
```diff
@@ -398,6 +398,41 @@ pub fn handle_find_element_elements_css(
...
+    reply: IpcSender<Result<Vec<String>, ()>>,
+) {
+    let node_ids = find_node_by_unique_id(documents, pipeline, element_id)
+        .ok_or(())
+        .and_then(|node| {
+            node.query_selector_all(DOMString::from("a"))
+                .map_err(|_| ())
+        })
+        .map(|nodes| {
+            nodes
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: To reduce a lot of the duplication in these commits, could we refactor the code so we have:
* a `all_matching_links` helper that accepts a root node, performs the query_selector_all operation, and returns a filtered iterator of unique ids
* a `first_matching_link` helper that calls `all_matching_links` and uses `take(1)` to return the first element of the filtered iterator

Using these building blocks, we should be able to write much more concise and less repetitive implementations for all of the new webdriver handlers.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/webdriver_handlers.rs (hunk 2 above)
Comment (georgeroman) [reply components/script/webdriver_handlers.rs]: I tried to return an iterator from `all_matching_links` but kept getting `Returns a value referencing data owned by the current function` compile-time errors. Instead, I simply collected the iterator and returned a vector of ids. Any thoughts on how to make it return an iterator?

**Comment by jdm** [comment]
@bors-servo r+
Much cleaner!

**Pull request merged**